test/composed-loop.cc

```diff
@@ -0,0 +1,66 @@
+#include "stout/composed-loop.h"
+
+#include <vector>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "stout/collect.h"
+#include "stout/iterate.h"
+#include "stout/map.h"
+#include "stout/range.h"
+#include "stout/terminal.h"
+#include "stout/then.h"
+
+using stout::eventuals::operator|;
+using stout::eventuals::Collect;
+using stout::eventuals::ComposedLoop;
+using stout::eventuals::Iterate;
+using stout::eventuals::Map;
+using stout::eventuals::Range;
+using stout::eventuals::Then;
+using testing::ElementsAre;
+
+TEST(ComposedLoop, TwoLevelLoop) {
+  auto s = [&]() {
+    return Range(2)
+        | ComposedLoop([](int x) { return Range(2); })
+        | Collect<std::vector<int>>();
+  };
+
+  EXPECT_THAT(*s(), ElementsAre(0, 1, 0, 1));
+}
+
+TEST(ComposedLoop, ThreeLevelLoop) {
+  auto s = [&]() {
+    return Range(2)
+        | ComposedLoop([](int x) { return Range(2); })
+        | ComposedLoop([](int x) { return Range(2); })
+        | Collect<std::vector<int>>();
+  };
+
+  EXPECT_THAT(*s(), ElementsAre(0, 1, 0, 1, 0, 1, 0, 1));
+}
+
+TEST(ComposedLoop, ComposedLoopMapped) {
+  auto s = [&]() {
+    return Range(2)
+        | ComposedLoop([](int x) { return Range(2); })
+        | Map(Then([](int x) { return x + 1; }))
+        | Collect<std::vector<int>>();
+  };
+
+  EXPECT_THAT(*s(), ElementsAre(1, 2, 1, 2));
+}
+
+TEST(ComposedLoop, ComposedLoopIterate) {
+  auto s = [&]() {
+    return Range(2)
+        | ComposedLoop([](int x) { 
+            std::vector<int> v = {1, 2, 3};
+          return Iterate(std::move(v)); })
+        | Map(Then([](int x) { return x + 1; }))
+        | Collect<std::vector<int>>();
+  };
+
+  EXPECT_THAT(*s(), ElementsAre(2, 3, 4, 2, 3, 4));
+}
```

Comment:
Since tests are documentation too, a test that demonstrates how the `x` being passed to `ComposedLoop` can be used would be nice.

Reply:
Also good to have: some test that does a ComposedLoop with something that isn't an `int`.

Reply:
Also good to have: some test that shows how to use a ComposedLoop _inside_ a ComposedLoop, so both `x` and `y` are accessible at the same time, for example when you want to do the equivalent of:
```c++
for (int x = 0; x < 2; x++) {
  for (int y = 10; y < 12; y++) {
    emit(x*y);
  }
}
```